services/web/server/src/simcore_service_webserver/projects/_project_models_rest.py

```diff
@@ -0,0 +1,398 @@
+        examples=[
+            "simcore/services/comp/sleeper",
+            "simcore/services/dynamic/3dviewer",
+            "simcore/services/frontend/file-picker",
+        ],
+    )
+    version: constr(
+        regex=r"^(0|[1-9]\d*)(\.(0|[1-9]\d*)){2}(-(0|[1-9]\d*|\d*[-a-zA-Z][-\da-zA-Z]*)(\.(0|[1-9]\d*|\d*[-a-zA-Z][-\da-zA-Z]*))*)?(\+[-\da-zA-Z]+(\.[-\da-zA-Z-]+)*)?$"
+    ) = Field(
+        ...,
+        description="semantic version number of the node",
+        examples=["1.0.0", "0.0.1"],
+    )
+    label: str = Field(
+        ..., description="The short name of the node", example=["JupyterLab"]
+    )
+    progress: Optional[confloat(ge=0.0, le=100.0)] = Field(
+        None, description="the node progress value"
+    )
+    thumbnail: Optional[AnyUrl] = Field(
+        None,
+        description="url of the latest screenshot of the node",
+        examples=["https://placeimg.com/171/96/tech/grayscale/?0.jpg"],
+    )
+    run_hash: Optional[Optional[str]] = Field(
+        None,
+        alias="runHash",
+        description="the hex digest of the resolved inputs +outputs hash at the time when the last outputs were generated",
+        examples=["a4337bc45a8fc544c03f52dc550cd6e1e87021bc896588bd79e901e2"],
+    )
+    inputs: Optional[
+        Dict[
+            constr(regex=r"^[-_a-zA-Z0-9]+$"),
+            Union[Optional[Union[int, bool, str, float]], Input, Input1, Input2],
+        ]
+    ] = Field(None, description="values of input properties")
+    input_access: Optional[
+        Dict[constr(regex=r"^[-_a-zA-Z0-9]+$"), InputAccess]
+    ] = Field(
+        None, alias="inputAccess", description="map with key - access level pairs"
+    )
+    input_nodes: Optional[List[UUID]] = Field(
+        None,
+        alias="inputNodes",
+        description="node IDs of where the node is connected to",
+        examples=["nodeUuid1", "nodeUuid2"],
+    )
+    outputs: Optional[
+        Dict[
+            constr(regex=r"^[-_a-zA-Z0-9]+$"),
+            Union[Optional[Union[int, bool, str, float]], Output, Output1],
+        ]
+    ] = {}
+    output_node: Optional[bool] = Field(None, alias="outputNode")
+    output_nodes: Optional[List[UUID]] = Field(
+        None,
+        alias="outputNodes",
+        description="Used in group-nodes. Node IDs of those connected to the output",
+        examples=["nodeUuid1", "nodeUuid2"],
+    )
+    parent: Optional[Optional[str]] = Field(
+        None,
+        description="Parent's (group-nodes') node ID s.",
+        examples=["nodeUuid1", "nodeUuid2"],
+    )
+    position: Optional[Position] = None
+    state: Optional[State] = Field(None, title="NodeState")
+
+
+class Position1(BaseModel):
+    class Config:
+        extra = Extra.forbid
+
+    x: int = Field(..., description="The x position", example=["12"])
+    y: int = Field(..., description="The y position", example=["15"])
+
+
+class Workbench1(BaseModel):
```

Comment:
ok, so if I get it right, these models are auto-generated with your new tool right?
are they in use?
and these suffixes? what do they mean?

Reply:
no, no ... :-(  I guess i did not explain it clearly in the description

This is created with [``datamodel-codegen``](https://github.com/koxudaxi/datamodel-code-generator)  from json-schema. See services/web/server/Makefile.

```
simcore_service_webserver/api/v0/schemas/project-v0.0.1.json --(datamodel-codegen)-->_project_models_rest.py
```
then this module is used as ``reference_cls`` to build all model variants of the ``Project`` resource. And it is here where we use ``models_factory.copy`` to build the model classes.

@sanderegg it is important that this idea gets through and we agree on it.

Reply:
I am also not 100 % sure I understand. So you suggest to parse all json-schemas that we have with the code-generator and that would give the baseline models (100% complete with all fields). Those will replace the various models.py in services and model library? And then from there you create specialized model classes (omitting for instance some fields). Is this done on the fly or are you running your tool over them and store the results in a static file? What happens if you want to add fields wrt to the baseline model?

Reply:
@mguidon here some clarifications

> I am also not 100 % sure I understand. So you suggest to parse all json-schemas that we have with the code-generator and that would give the baseline models (100% complete with all fields). Those will replace the various models.py in services and model library? 

Basically yes. This aligns as well with "specifications driven design". We use as "source of truth" the language-agnostic json-schemas (or OAS) specs . Then we "translate" them into pydantic models by auto-generating them into reference models (this is how i denote what you call "baseline models"). BTW we could actually do this also at runtime instead of hard-coding it.

> And then from there you create specialized model classes (omitting for instance some fields). Is this done on the fly or are you running your tool over them and store the results in a static file? 

This is done at runtime (IMO this does not has an overhead) but it could be relatively easy to extend it to produce "hard-coded" classes. This would definitively make more clear "the final model class" .  The only drawback with that approach is that we need to re-generate and commit every time there is a change and my experience tells me that we are not very good at that :-) 

> What happens if you want to add fields wrt to the baseline model?

You can use the generated class as a base class as well (see usage examples in ``packages/models-library/tests/test_utils_models_factory.py``)